src/main/kotlin/com/jetbrains/snakecharm/codeInsight/SnakemakeParamsInShellReferenceContributor.kt

```diff
@@ -0,0 +1,48 @@
+package com.jetbrains.snakecharm.codeInsight
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.patterns.PlatformPatterns
+import com.intellij.psi.*
+import com.intellij.psi.util.parentOfType
+import com.intellij.util.ProcessingContext
+import com.jetbrains.python.psi.PyStringLiteralExpression
+import com.jetbrains.snakecharm.lang.SnakemakeLanguageDialect
+import com.jetbrains.snakecharm.lang.psi.SMKParamsReference
+import com.jetbrains.snakecharm.lang.psi.SMKRuleParameterListStatement
+import java.util.regex.Pattern
+
+class SnakemakeParamsInShellReferenceContributor : PsiReferenceContributor() {
+    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
+        registrar.registerReferenceProvider(
+                PlatformPatterns.psiElement(),
+                object : PsiReferenceProvider() {
+                    private val paramsPattern = Pattern.compile("\\{params\\.(.+?)}")
+
+                    override fun getReferencesByElement(
+                            element: PsiElement,
+                            context: ProcessingContext
+                    ): Array<PsiReference> {
+                        if (element.containingFile.language != SnakemakeLanguageDialect) {
+                            return emptyArray()
+                        }
+
+                        val paramReferences = mutableListOf<PsiReference>()
+
+                        val isShellCommand = element
```

Comment:
See `RuleSectionKeywordsProvider`, you can move language check and `inside SMKRuleParameterListStatement` check into pattern used in `registrar.registerReferenceProvider`

Reply:
I can remove the `is PyStringLiteral` check but I haven't been able to come up with the correct pattern for checking if we are inside the _shell_ section, not just any rule section: the element we're looking at is a string literal, its parent is a `PyArgumentList`, and its parent is `SMKRuleParameterListStatement`, whose name, according to its `getName()` method, is "shell" but who does not match the `PlatformPatterns.psiElement(SMKRuleParameterListStatement::class.java).withName("shell")` pattern somehow. So, yes, I can modify the pattern used in `registrar.registerReferenceProvider`, but I still don't see how I can move the `isShellCommand` check to the pattern because the pattern that, I think, should work, doesn't work, and I don't see how to form this pattern in another way.

Reply:
I meant only check that element is inside required parent and in file with proper language, not section name

Reply:
That means autocompletion/resolve for params would be available not only in shell section, but in all other rule sections as well. Should I add corresponding tests?

Reply:
I fixed the pattern but the safe call in `.getParentOfType(element, SMKRuleParameterListStatement::class.java)?.section` still can't be removed, compilation error.

Reply:
What kind of compilation error?